metricbeat/module/prometheus/query/query.go

```diff
@@ -0,0 +1,119 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package query
+
+import (
+	"fmt"
+	"io/ioutil"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/metricbeat/helper"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	"github.com/elastic/beats/v7/metricbeat/mb/parse"
+)
+
+const (
+	defaultScheme = "http"
+)
+
+var (
+	hostParser = parse.URLHostParserBuilder{
+		DefaultScheme: defaultScheme,
+	}.Build()
+)
+
+func init() {
+	mb.Registry.MustAddMetricSet("prometheus", "query", New,
+		mb.WithHostParser(hostParser),
+	)
+}
+
+// MetricSet type defines all fields of the MetricSet for Prometheus Query
+type MetricSet struct {
+	mb.BaseMetricSet
+	http    *helper.HTTP
+	queries []QueryConfig
+	baseURL string
+}
+
+// New create a new instance of the MetricSet
+// Part of new is also setting up the configuration by processing additional
+// configuration entries if needed.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	config := defaultConfig()
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	http, err := helper.NewHTTP(base)
+	if err != nil {
+		return nil, err
+	}
+	return &MetricSet{
+		BaseMetricSet: base,
+		http:          http,
+		queries:       config.Queries,
+		baseURL:       http.GetURI(),
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(reporter mb.ReporterV2) error {
+	for _, pathConfig := range m.queries {
+		url := m.getURL(pathConfig.Path, pathConfig.QueryParams)
+		m.http.SetURI(url)
+		response, err := m.http.FetchResponse()
+		if err != nil {
+			msg := fmt.Sprintf("unable to fetch data from prometheus endpoint: %v", pathConfig.Path)
+			m.Logger().Debugf("%v: %v", msg, err)
+			reporter.Error(errors.Wrap(err, msg))
+			continue
+		}
+		defer func() {
+			if err := response.Body.Close(); err != nil {
+				m.Logger().Debug("error closing http body")
+			}
+		}()
+
+		body, err := ioutil.ReadAll(response.Body)
```

Comment:
Avoid `ioutil.ReadAll`, use `json.NewDecoder(response.Body).Decode(&yourStruct)`. `response.Body` is not being closed, btw

Reply:
`response.Body` is being closed 2 lines above :), no? 

As far as the `ioutil.ReadAll`, it is required since we don't know beforehand the type of response and the parsing happens later on in `data.go`. Otherwise we should go with a try-error approach on `json.NewDecoder(response.Body).Decode(&yourStruct)` attempting to Decode with all the possible ResponseStructs. wdyt?

Reply:
Ouch, yes, response is closed, sorry.

Without entering into details, the problem of `ReadAll` is that it creates a huge byte array that is then "duplicated" when unmashalling. But while "following" the code, I see more things to rework.

I'm going step by step writing while "thinking loud" :smile: 
* `response.Body` can be passed into `parseResponse` because it is not necessary before.
* `parseResponse` returns events so it should be called `getEventsFromResponseBody` maybe
* By the rule of clean code of "a function must do only one thing, but do it very well" we must have a `get[myStruct]FromResponseBody` function and a `createEventsFrom[myStruct]`. You can argument that we are trying to avoid more `memory allocations` however, I prefer code readability and proper because beats are executed only once every X seconds.
* `myStruct` will help unit testing of `get[myStruct]FromResponseBody` and `createEventsFrom[myStruct]` separatedly.
* The approach in `parseResponse` of fully parse the entire JSON to check its schema should be avoided. A library like https://github.com/buger/jsonparser could be used to avoid this but it will introduce a new dependency in beats, in any case it will finally require the use if `ReadAll`, so we can forget about the `json.NewDecoder` for this metricset anyways.
*